sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexingBufferedAsyncSender.java

```diff
@@ -15,66 +14,93 @@
+    private final Object actionsMutex = new Object();
+    private List<TryTrackingIndexAction<T>> actions = new ArrayList<>();
 
     private final AtomicReference<TimerTask> flushTask = new AtomicReference<>();
 
-    SearchBatchAsyncClient(SearchAsyncClient client, Boolean autoFlush, Duration flushWindow,
-        Integer batchSize, IndexingHook indexingHook) {
-        ClientLogger logger = new ClientLogger(SearchBatchAsyncClient.class);
-
-        if (batchSize != null && batchSize < 1) {
-            throw logger.logExceptionAsError(new IllegalArgumentException("'batchSize' cannot be less than one."));
-        }
+    SearchIndexingBufferedAsyncSender(SearchAsyncClient client, SearchIndexingBufferedSenderOptions<T> options) {
+        SearchIndexingBufferedSenderOptions<T> buildOptions = (options == null)
+            ? new SearchIndexingBufferedSenderOptions<>()
+            : options;
 
         this.client = client;
-        this.autoFlush = autoFlush == null || autoFlush;
-
-        if (flushWindow == null) {
-            this.flushWindowMillis = DEFAULT_FLUSH_WINDOW;
-        } else if (flushWindow.isZero() || flushWindow.isNegative()) {
-            this.flushWindowMillis = 0;
-        } else {
-            this.flushWindowMillis = flushWindow.toMillis();
-        }
+        this.autoFlush = buildOptions.getAutoFlush();
+        this.flushWindowMillis = Math.max(0, buildOptions.getFlushWindow().toMillis());
+        this.batchSize = buildOptions.getBatchSize();
+        this.documentTryLimit = buildOptions.getDocumentTryLimit();
+
+        this.onActionAddedConsumer = (action) -> {
+            if (buildOptions.getOnActionAdded() != null) {
+                buildOptions.getOnActionAdded().accept(action);
+            }
+        };
+
+        this.onActionSucceededConsumer = (action) -> {
+            if (buildOptions.getOnActionSucceeded() != null) {
+                buildOptions.getOnActionSucceeded().accept(action);
+            }
+        };
+
+        this.onActionErrorBiConsumer = (action, throwable) -> {
+            if (buildOptions.getOnActionError() != null) {
+                buildOptions.getOnActionError().accept(action, throwable);
+            }
+        };
+
+        this.onActionRemovedConsumer = (action) -> {
+            if (buildOptions.getOnActionRemoved() != null) {
+                buildOptions.getOnActionRemoved().accept(action);
+            }
+        };
+
+        this.documentKeyRetriever = (buildOptions.getDocumentKeyRetriever() != null)
+            ? buildOptions.getDocumentKeyRetriever()
+            : buildDocumentKeyRetriever(buildOptions.getClass().getGenericSuperclass());
 
-        this.batchSize = (batchSize == null) ? DEFAULT_BATCH_SIZE : batchSize;
-        this.indexingHook = indexingHook;
         this.autoFlushTimer = (this.autoFlush) ? new Timer(true) : null;
     }
 
+    private Function<T, String> buildDocumentKeyRetriever(Type clazz) {
+        return null;
```

Comment:
Need to figure out how to pass generic information here cleanly. Currently, this has the issue where the type is erased at this point where `T` resolved to `Object`. .NET uses `FieldBuilder` functionality or retrieving the index from the Search service to generate an implicit `documentKeyRetriever` instead of explicitly requiring it to be configured all the time. So, either a `Class<T>` will need to be passed somewhere or Java will need to require explicit configuration of the `documentKeyRetriever`.

Reply:
@JonathanGiles @srnagar @tg-msft what are your thoughts on this?

Reply:
Given this will go out for another preview Java will be making this a required configuration.